cpp/autosar/src/rules/A16-0-1/PreProcessorShallOnlyBeUsedForCertainDirectivesPatterns.ql

```diff
@@ -40,10 +53,18 @@ pragma[noinline]
 predicate isPreprocConditionalRange(
   PreprocessorBranch pb, string filepath, int startLine, int endLine
 ) {
-  exists(PreprocessorEndif end | pb.getEndIf() = end |
-    isPreprocFileAndLine(pb, filepath, startLine) and
-    isPreprocFileAndLine(end, filepath, endLine)
-  )
+  //the range of an if with an elif ends at the elif to avoid reporting things twice
+  if exists(PreprocessorElif elif | elif.getIf() = pb)
+  then
+    exists(PreprocessorElif end | end.getIf() = pb |
+      isPreprocFileAndLine(pb, filepath, startLine) and
+      isPreprocFileAndLine(end, filepath, endLine)
```

Comment:
I think we want the least range by looking at the closest `elseif` in case of multiple ones. Something like (not tested):

```suggestion
    isPreprocFileAndLine(pb, filepath, startLine) and
    endLine = min(int elifEndLine, PreProcessorElif end | isPreprocFileAndLine(end, filepath, elifEndLine) | elifEndLine)
```

Reply:
great catch , this didnt quite associate the end item to the one above, but ended up simplifying the whole thing immensely!